For `st.image` return "GIF", not "PNG", if PIL image is a GIF

snehankekre opened the pull request:
## 📚 Context

#5418 reports a regression in `st.image` for Streamlit 1.13.0 (possibly introduced by #5383 or #5287), whereby passing the path of a local GIF file results in the image being converted to a static PNG image.

### Why does this happen?

- When `st.image` is passed the path of a local GIF file, the following is executed because the path is a string: https://github.com/streamlit/streamlit/blob/344e72818a5d27fe54932604bcb658b79ba259fa/lib/streamlit/elements/image.py#L483-L497 
- As `image` is a string, and not an svg, `image` is passed to `image_to_url`: https://github.com/streamlit/streamlit/blob/344e72818a5d27fe54932604bcb658b79ba259fa/lib/streamlit/elements/image.py#L314-L327
- As the string is not a url, it is opened as a file, and the resultant bytes are passed to `_validate_image_format_string`: https://github.com/streamlit/streamlit/blob/344e72818a5d27fe54932604bcb658b79ba259fa/lib/streamlit/elements/image.py#L377-L378
- ... where the bytes are read as a PIL image and, as a result of the PIL image having an alpha channel, it is cast to "PNG": https://github.com/streamlit/streamlit/blob/344e72818a5d27fe54932604bcb658b79ba259fa/lib/streamlit/elements/image.py#L185-L191

- What kind of change does this PR introduce?

  - [x] Bugfix: st.image regression in v1.13.0

## 🧠 Description of Changes

- Adds a check to `_validate_image_format_string()` to identify if the PIL image is a GIF and, if so, return "GIF"

  - [x] This is a visible (user-facing) change


## 🧪 Testing Done

- [x] Added/Updated unit tests
- [x] Added/Updated e2e tests

## 🌐 References


- **Issue**: Closes #5418 

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

snehankekre committed:
Return "GIF", not "PNG", if PIL image is a GIF

snehankekre committed:
Add "GIF" type to ImageFormat

snehankekre committed:
Test for GIF using image.format instead of GifImagePlugin.GifImageFile

lukasmasuch commented:
> TODO: I could use some help writing tests for GIFs 😅

e2e tests are probably complicated for GIFs, but you could take a look into [`image_test.py` ](https://github.com/streamlit/streamlit/blob/develop/lib/tests/streamlit/elements/image_test.py) to add some unit tests for the python function to test the gif handling. Currently, it doesn't contain anything related to gifs. 

And maybe you can add an e2e test that just checks if the image path in the `img` property uses the `.gif` extension. Currently, the image is send as a `.png`.

snehankekre committed:
Add unit tests for st.image GIF handling

snehankekre committed:
Update st.image e2e tests for GIFs

snehankekre committed:
Take a better imaged snapshot of GIF with caption

snehankekre committed:
Fix index: Take a snapshot of the right image

snehankekre committed:
Update st.image e2e snapshot for GIFs

snehankekre commented:
Hi @LukasMasuch, thanks for the pointers. I've added unit tests for GIF handling on cf48552, and also added e2e tests for GIFs in subsequent commits. PTAL 😄 For the e2e tests, I generated a simple, deterministic gif to reduce complexity and flakiness.

LMK what you think of the tests, and if they are sufficient and meet our standards or don't.

lukasmasuch commented on e2e/specs/st_image.spec.js:
```diff
@@ -103,4 +105,24 @@ describe("st.image", () => {
+  it("displays a GIF image and a caption together", () => {
+    cy.get(".element-container [data-testid='stImage']")
+      .eq(15)
+      .matchImageSnapshot("gif-with-caption");
```
Does the gif always show the same content, or is there a possibility for flakiness here?

snehankekre replied:
I've made it so that the gif always shows the same content, no randomness involved: 
![gif](https://user-images.githubusercontent.com/20672874/193065212-5af99578-e6ef-4dea-a05d-5d6e37a048b8.gif)

lukasmasuch replied:
But it is still moving and the snapshot is only taking a still image from some kind of random time. So it could take the snapshot at different stages of the gif, or not?

snehankekre replied:
Another solid point! You're right, the test could have gotten flaky. I was under the false impression that gifs were considered animations and that cypress would [waitForAnimations](https://docs.cypress.io/guides/references/configuration#Actionability) to finish animating before executing commands.

Learned the hard way that gifs aren't animations (they don't animate DOM elements) and there's no easy way in JS to check if a gif has stopped.

What I've done in follow up commits is:
- programmed the GIF to stop moving in 3200 ms.
- use cypress to get the DOM element
- wait for 4000 ms (guarantees that snapshot will be taken once gif stops)
- take the snapshot

While I recognize that using `cy.wait(<arbitrary-period>)` is an [anti-patten](https://docs.cypress.io/api/commands/wait#Time), I've been unable to find another way to check if the gif has stopped moving. In an ideal world, when the gif stops, we would hope that some HTML element or CSS property changed, so we can wait for that property instead of a time period, but that's not the case.

With my current implementation, the snapshot taken will always be the same. i.e. at the end stage of the gif; avoiding flakiness. 
![image](https://user-images.githubusercontent.com/20672874/193244579-967e5a85-cb97-4f6b-8d6e-43d3ca95c043.png)

lukasmasuch replied:
Thanks for updating. I think this is a valid solution. Unfortunately, 4 seconds of waiting is too long. It would slow down our CI tests. I would propose either getting it down to under 1000ms (using fewer frames) or we leave out this snapshot test for now. Maybe add a `frames` parameter to the `create_gif` function to configure the number of frames, and use only one frame for the snapshot gif?

snehankekre replied:
Thanks Lukas! I'll get it down under 1000ms or use only one frame for the snapshot. Give me a day or two to fix the tests, I'm currently OOO

lukasmasuch replied:
Sounds good :)

snehankekre replied:
@LukasMasuch I adopted your idea of including a `frames` parameter, and for the snapshot we create a gif with only one frame and take the snapshot. There is 0 movement in the gif, so no chance of flakiness. Also, removed the 4 second delay.

snehankekre replied:
The snapshot will always look like this 
![image](https://user-images.githubusercontent.com/20672874/194481723-7f1e3cf9-8f28-4da0-a443-a4282f03037b.png)

lukasmasuch commented on lib/streamlit/elements/image.py:
```diff
@@ -164,15 +168,22 @@ def _image_may_have_alpha_channel(image: PILImage) -> bool:
 
 
+def _image_is_gif(image: PILImage) -> bool:
+    if image.format == "GIF":
```
nit: You can simplify this to:
```python
return image.format == "GIF"
```

snehankekre replied:
Good catch! Fixed on a954fc5.

lukasmasuch replied:
Ahh, interesting. This seems to be a type violation. Probably works if you wrap it into a `bool(image.format == "GIF")`

snehankekre replied:
I reverted the change because `return image.format == "GIF"` returns type `Any`, not `bool`, and mypy complains

snehankekre replied:
> Probably works if you wrap it into a bool(image.format == "GIF")

That worked!

lukasmasuch commented on lib/streamlit/elements/image.py:
```diff
@@ -50,12 +52,14 @@
 UseColumnWith: TypeAlias = Optional[Union[Literal["auto", "always", "never"], bool]]
 Channels: TypeAlias = Literal["RGB", "BGR"]
-ImageFormat: TypeAlias = Literal["JPEG", "PNG"]
+ImageFormat: TypeAlias = Literal["JPEG", "PNG", "GIF"]
```
Since this also adds `gif` as a supported argument to `output_format` we might need to add it to the docstring as well. Can I load a `png` image and specify `gif` as output format and it transforms it into a `gif`?

snehankekre replied:
That's a good point! I added the "GIF" type to `ImageFormat` only to avoid:

```sh
lib/streamlit/elements/image.py: note: In function "_validate_image_format_string":
lib/streamlit/elements/image.py:202:16: error: Incompatible return value type (got "Literal['GIF']", expected "Literal['JPEG', 'PNG']")  [return-value]
```

If you load any non-gif image (png, jpg, svg) and specify `output_format="GIF"`, it is not transformed into a gif. It follows the st.image logic and returns the same type as input or "PNG" or "JPG".

lukasmasuch replied:
Alright, probably we don't need to add any warning here that the argument is ignored since we don't do this for other situations as well (e.g. using an SVG as input with output_format=PNG).

lukasmasuch commented:
Great work 💯  Have you tested if this also works for `GIFs` loaded from a local file?

snehankekre committed:
Simplify _image_is_gif to one-liner

snehankekre committed:
Return Any instead of bool in _image_is_gif

snehankekre committed:
Revert changes to _image_is_gif to return bool

snehankekre commented:
> Have you tested if this also works for GIFs loaded from a local file?

Yes, I've tested that locally. In the unit tests and e2e tests rather than write a GIF file to CI container, I write a GIF to a buffer in memory and pass that to `st.image`. In the st.image logic, if you pass a path to a local GIF file, the file is opened and read into bytes. So, in effect the unit+e2e tests are also reading from a local file

snehankekre committed:
Make _image_is_gif return bool

snehankekre committed:
Make _image_is_gif return bool

snehankekre committed:
Don't make GIF loop in e2e test

snehankekre committed:
Try waiting for 4s before taking snapshot of GIF

snehankekre committed:
Delete offending snapshots

snehankekre committed:
Use more specific selector to take snapshot of GIF

snehankekre committed:
Describe what GIF does in test comments

snehankekre committed:
Update gif-with-caption snapshot for st.image e2e

snehankekre committed:
Merge branch 'develop' into st-image-local-gif

snehankekre committed:
Create GIF with 1 frame to take snapshot of

snehankekre committed:
Remove 4s delay from st.image e2e for GIFs

snehankekre committed:
Delete obsolete snapshot

snehankekre committed:
Update gif-with-caption snapshot for st.image e2e

lukasmasuch reviewed (approved):
LGTM 👍
